enhancements/storage/csi-ebs-operator.md

```diff
@@ -0,0 +1,142 @@
+
+## Summary
+
+In the past storage drivers have been provided either in-tree or via external provisioners. With the move to advocating CSI as the way to deliver storage drivers, Red Hat needs to move the in-tree drivers (and others that our customers need) to use CSI. With the focus on operators in Red Hat OCP 4.x these CSI drivers should be installed in the form of an Operator.
+
+## Motivation
+
+* In-tree drivers will be removed from Kubernetes, we need to continue to make the drivers available and CSI is the way to do this.
+* AWS is a key cloud provider for OpenShift and use of EBS has been supported in both OpenShift 3.x and 4.x, we need to move this driver to be CSI, provided by an operator.
+* CSI drivers (incl. AWS EBS) will enable new storage features, not possible with in-tree volume plugins, like snapshots and volume cloning.
+
+### Goals
+
+* Create an operator to install the AWS EBS CSI driver.
+* Publish the operator on OperatorHub.
+* Package and ship downstream image of AWS EBS CSI driver (upstream: https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
+
+### Non-Goals
+
+* Driver creation (upstream: https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
+
+## Proposal
+
+1. OCP ships a new operator called aws-ebs-csi-driver-operator. This operator:
+   * Installs the AWS EBS CSI driver.
+	 * It creates a new Deployment that runs the controller pieces of the driver (provisioner, attacher, resizer, etc.).
+	 * It creates a new DaemonSet that runs the CSI driver on the worker nodes.
+   * Is installed by OLM once the user opts-in.
+   * Runs in namespace `openshift-aws-ebs-csi-driver-operator`, and its operands run in namespace `openshift-aws-ebs-csi-driver`.
+
+### API
+
+While the operator does not need any special cluster-specific configuration, following API is provided. Explicitly, the operator uses `status.observedGeneration` and `status.generations` to track changes in dependent objects.
+
+Package name: `github.com/openshift/aws-ebs-csi-driver-operator/pkg/apis/operator/v1alpha1`
+
+```go
+// EBSCSIDriver is a specification for a EBSCSIDriver resource
+type EBSCSIDriver struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   EBSCSIDriverSpec   `json:"spec"`
+	Status EBSCSIDriverStatus `json:"status"`
+}
+
+// EBSCSIDriverSpec is the spec for a EBSCSIDriver resource
+type EBSCSIDriverSpec struct {
+	operatorv1.OperatorSpec `json:",inline"`
+}
+
+// EBSCSIDriverStatus is the status for a EBSCSIDriver resource
+type EBSCSIDriverStatus struct {
+	operatorv1.OperatorStatus `json:",inline"`
+}
+
+// EBSCSIDriverList is a list of EBSCSIDriver resources
+type EBSCSIDriverList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []EBSCSIDriver `json:"items"`
+}
+```
+
+The CRD and its `cluster` instance is created by OLM from the operator manifest.
```

Comment:
@gnufied, can we have `cluster`  instance created by OLM? That would be awesome. Plus having a webhook to prevent other instances to be created.

Reply:
Discussed on a call, OLM cannot create the CR. The operator should create the CR by itself if it does not exist.

Reply:
The cluster instance of CR can defenitely be created by the OLM/operator when operator is installed. I dunno about webhooks. Haven't seen any operators that ship with one.

Reply:
Thinking again about the default CR. If the CR is going to be created by OLM when the operator is installed, there can be only one hardcoded CR there. Maybe we need some better logic:

* On a cluster with AWS cloud provider configured, there *should* be a default CR created and the operator gets AWS credentials through cloud-credential-operator.
* On a "bare metal" cluster on AWS VMs, where admin does not have cloud provider configured, yet wants to use AWS EBS volumes (we've seen a support ticket for this), there *should be no CR created*, as admin must provide AWS credentials manually and refer them from the CR.

Reply:
scratch the above, OLM cannot create CR on startup

Reply:
So the operator itself should create the only CR when it starts && the cluster has cloud provider configured? And not create the CR otherwise? That sounds odd...

Reply:
As we discussed on a call yesterday, for 4.5 the CR will be installed by the user.